packages/auth/__tests__/auth.int.jest.js

```diff
@@ -206,6 +205,151 @@ test('sign in with tokenType APP default tokenType', async () => {
+    expect(resultSignIn.signInResult.data.signIn.phrase).toBeTruthy()
+    expect(resultSignIn.signInResult.errors).toBeFalsy()
+
+    const resultAuthorize = await authorizeSession({
+      email: Users.Unverified.email,
+      tokens: [{ type: tokenType, payload: resultSignIn.payload }],
+      apolloFetch: global.instance.createApolloFetch()
+    })
+
+    expect(resultAuthorize.data).toBeFalsy()
+  })
+
+  test('fails if amount of simultaniously valid tokens is exceeded', async () => {
+    for (let i = 0; i < MAX_VALID_TOKENS; i++) {
+      await signIn({
+        user: Users.Unverified,
+        tokenType,
+        skipAuthorization: true
+      })
+    }
+
+    const resultSignIn = await signIn({ user: Users.Unverified, tokenType, skipAuthorization: true })
+    expect(resultSignIn.signInResult.errors).toBeTruthy()
+  })
+
+  test('fails if amount of simultaniously valid tokens in multiple sessions is exceeded', async () => {
+    for (let i = 0; i < MAX_VALID_TOKENS; i++) {
+      await signIn({
+        user: Users.Unverified,
+        tokenType,
+        skipAuthorization: true,
+        newCookieStore: true
+      })
+    }
+
+    const resultSignIn = await signIn({
+      user: Users.Unverified,
+      tokenType,
+      skipAuthorization: true,
+      apolloFetch: global.instance.createApolloFetch()
+    })
+    expect(resultSignIn.signInResult.errors).toBeTruthy()
+  })
+
+  test('authorize after some valid tokens expired', async () => {
+    for (let i = 0; i < 5; i++) {
+      await signIn({
+        user: Users.Unverified,
+        tokenType,
+        skipAuthorization: true
+      })
+    }
+
+    const tokens = (await pgDatabase().public.tokens.find(
+      { email: Users.Unverified.email }
+    ))
+
+    await pgDatabase().public.tokens.update(
+      { id: tokens.map(token => token.id)[0] },
+      { expiresAt: new Date() }
+    )
+
+    const resultSignIn = await signIn({ user: Users.Unverified, tokenType, skipAuthorization: true })
+    expect(resultSignIn.signInResult.data.signIn.phrase).toBeTruthy()
+    expect(resultSignIn.signInResult.errors).toBeFalsy()
+  })
+})
+
+describe('authorizeSession', () => {
+  test('rate limit attempts within session', async () => {
+    const resultSignIn = await signIn({
+      user: Users.Unverified,
+      tokenType: 'EMAIL_TOKEN',
+      skipAuthorization: true
+    })
+
+    const { apolloFetch } = resultSignIn
+
+    for (let i = 0; i < 10; i++) {
+      const resultAuthorize = await authorizeSession({
+        email: Users.Unverified.email,
+        tokens: [{ type: 'EMAIL_TOKEN', payload: `attempt-${i}` }],
+        apolloFetch
+      })
+
+      expect(resultAuthorize.data).toBeFalsy()
+    }
+
+    const resultAuthorize = await authorizeSession({
+      email: Users.Unverified.email,
+      tokens: [{ type: 'EMAIL_TOKEN', payload: resultSignIn.payload }],
+      apolloFetch
+    })
+
+    expect(resultAuthorize.data).toBeFalsy()
```

Comment:
It's not clear to me what the end state of this test is. the last attempt is also denied, or does it go through? Also how does one recover from too many failed attempts to authrorize a session?

Reply:
Last attempt won't go through, it does "rate limit attempts within session" (named test that way). Error returned is a generic `api/token/invalid` error message to not give away anything, but could assert that so it is more explicit.

A user can not recover from a rate limited session by an action, but will have to wait until either of the tokens linked to rate limited sessions is expired. [Expired tokens won't be taken into account when auditing authorize attempts.](https://github.com/orbiting/backends/blob/2d0a086c2b396afef9b2217c40dc9ed8394eb2b7/packages/auth/lib/Users.js#L287-L298)

Reply:
I see. The session is not found in this query if **all** the tokens have expired, thus it's not evaluated for `authorizeAttempts`. That makes sense. As this ratelimiter is not scoped to EMAIL_CODE I would go for an explicit error like "Anzahl Anmelde versuche überschritten, bitte warten Sie xx minuten." So that the careteam and we can see if people bump into this.

Reply:
Adopting.